orc8r/cloud/go/services/certifier/analytics/calculations/lifespan.go

```diff
@@ -0,0 +1,105 @@
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package calculations
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"io/ioutil"
+	"math"
+	"strings"
+	"time"
+
+	"magma/orc8r/cloud/go/services/analytics/calculations"
+	"magma/orc8r/cloud/go/services/analytics/protos"
+	"magma/orc8r/cloud/go/services/analytics/query_api"
+	"magma/orc8r/lib/go/metrics"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type CertLifespanCalculation struct {
+	calculations.BaseCalculation
+}
+
+const (
+	DefaultCertsDirectory = "/var/opt/magma/certs/"
+	RootCert              = "rootCA.pem"
+	AdminCert             = "admin_operator.pem"
+	ControllerCert        = "controller.crt"
+	CertifierCert         = "certifier.pem"
+	FluentdCert           = "fluentd.pem"
+)
+
+func (x *CertLifespanCalculation) Calculate(prometheusClient query_api.PrometheusAPI) ([]*protos.CalculationResult, error) {
+	glog.V(2).Infof("Calculating %s", metrics.CertLifeHoursMetric)
+
+	var results []*protos.CalculationResult
+
+	metricConfig, ok := x.AnalyticsConfig.Metrics[metrics.CertLifeHoursMetric]
+	if !ok {
+		glog.Errorf("%s metric not found in metric config", metrics.CertLifeHoursMetric)
+		return results, nil
+	}
+
+	var certs = [...]string{RootCert, AdminCert, ControllerCert, CertifierCert, FluentdCert}
+
+	for _, certName := range certs {
+		result, err := calculateCertLifespanHours(certName, metricConfig.Labels)
+		if err != nil {
+			glog.Errorf("Could not get lifespan for cert %s: %s", certName, err)
+			continue
+		}
+		results = append(results, result)
+	}
+
+	return results, nil
+}
+
+func calculateCertLifespanHours(certName string, metricConfigLabels map[string]string) (*protos.CalculationResult, error) {
+	dat, err := getCertBytes(certName)
+	if err != nil {
+		return nil, err
+	}
+	cert, err := x509.ParseCertificate(dat)
+	if err != nil {
+		return nil, err
+	}
+	// Hours left
+	hoursLeft := math.Floor(cert.NotAfter.Sub(time.Now()).Hours())
+	labels := prometheus.Labels{
+		metrics.CertNameLabel: certName,
+	}
+	labels = calculations.CombineLabels(labels, metricConfigLabels)
+	result := calculations.NewResult(hoursLeft, metrics.CertLifeHoursMetric, labels)
+	glog.V(2).Infof("Calculated metric %s for %s: %f", metrics.CertLifeHoursMetric, certName, hoursLeft)
+	return result, nil
+}
+
+func getCertBytes(certName string) ([]byte, error) {
+	dat, err := ioutil.ReadFile(DefaultCertsDirectory + certName)
+	if err != nil {
+		return nil, err
+	}
+	if strings.HasSuffix(certName, ".pem") || strings.HasSuffix(certName, ".crt") {
```

Comment:
If file name isn't pem or crt should we fail here ?

Reply:
Either way, use an early return rather than nested if block

Reply:
pem and crt files need to get decoded, but other files can be directly parsed as a certificate through the x509 package